compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -2831,11 +2831,31 @@ fn add_gcc_ld_path(cmd: &mut dyn Linker, sess: &Session, flavor: LinkerFlavor) {
                     // Implement the "linker flavor" part of -Zgcc-ld
                     // by asking cc to use some kind of lld.
                     cmd.arg("-fuse-ld=lld");
+
                     if !flavor.is_gnu() {
                         // Tell clang to use a non-default LLD flavor.
                         // Gcc doesn't understand the target option, but we currently assume
                         // that gcc is not used for Apple and Wasm targets (#97402).
-                        cmd.arg(format!("--target={}", sess.target.llvm_target));
+                        //
+                        // Note that we don't want to do that by default on macOS: e.g. passing a
+                        // 10.7 target to LLVM works, but not to recent versions of clang/macOS, as
+                        // shown in issue #101653 and the discussion in PR #101792.
+                        //
+                        // It could be required in some cases of cross-compiling with
+                        // `-Zgcc-ld=lld`, but this is generally unspecified, and we don't know
+                        // which specific versions of clang, macOS SDK, host and target OS
+                        // combinations impact us here.
+                        //
+                        // So we do a simple first-approximation until we know more of what the
+                        // Apple targets require (and which would be handled prior to hitting this
+                        // `-Zgcc-ld=lld` codepath anyway), but the expectation is that until then
+                        // this should be manually passed if needed.
+                        let is_targeting_wasm = matches!(flavor, LinkerFlavor::WasmLld(..));
```

Comment:
Not sure why WASM is special cased here, it is already covered by the linker flavor condition.

Reply:
It was less about special casing WASM than having to pick a bool between wasm and darwin, as we have to add `--target` on wasm, and not on darwin. 

So the choice was between `if wasm or different_linker_flavor` and `if !darwin or different_linker_flavor` and the first one seemed clearer to read but I can flip it to the second if you prefer ?

Reply:
I mean `different_linker_flavor` should always be true for WASM, so `wasm or different_linker_flavor == different_linker_flavor`.

Reply:
Done